src/clj/rems/auth/oidc.clj

```diff
@@ -23,25 +27,89 @@
+           (set (get-userid-attributes {:old_sub "alice"
+                                        :elixirId "elixir-alice"
+                                        :name "Alice Applicant"}))))))
+
+(defn get-userid [id-data]
+  (let [userids (vals (get-userid-attributes id-data))]
+    (first userids)))
 
 (deftest test-get-userid
-  (with-redefs [env {:oidc-userid-attribute "atr"}]
+  (with-redefs [env {:oidc-userid-attributes [{:attribute "atr"}]}]
     (is (nil? (get-userid {:sub "123"})))
     (is (= "456" (get-userid {:sub "123" :atr "456"}))))
-  (with-redefs [env {:oidc-userid-attribute ["atr"]}]
-    (is (nil? (get-userid {:sub "123"})))
-    (is (= "456" (get-userid {:sub "123" :atr "456"}))))
-  (with-redefs [env {:oidc-userid-attribute ["atr" "sub" "fallback"]}]
+  (with-redefs [env {:oidc-userid-attributes [{:attribute "atr"}
+                                              {:attribute "sub"}
+                                              {:attribute "fallback"}]}]
     (is (= "123" (get-userid {:sub "123"})))
     (is (= "456" (get-userid {:sub "123" :atr "456"})))
     (is (= "456" (get-userid {:sub "123" :atr "456" :fallback "78"})))
-    (is (= "78" (get-userid {:fallback "78"})))))
+    (is (= "78" (get-userid {:fallback "78"}))))
+  (with-redefs [env {:oidc-userid-attributes [{:attribute "atr" :rename "elixirId"}
+                                              {:attribute "sub"}]}]
+    (is (= "elixir-alice" (get-userid {:elixirId "elixir-alice" :sub "123"})))))
+
+(defn rename-userid-attributes [id-data]
+  (let [attrs (getx env :oidc-userid-attributes)]
+    (->> attrs
+         (filter :rename)
+         (map (partial map-vals keyword))
+         (reduce #(replace-key %1 (:attribute %2) (:rename %2)) id-data))))
+
+(deftest test-rename-user-attributes
+  (with-redefs [env {:oidc-userid-attributes [{:attribute "sub" :rename "elixirId"}
+                                              {:attribute "old_sub"}]}]
+    (is (nil? (rename-userid-attributes nil)))
+    (is (= #{[:old_sub "alice"]
+             [:elixirId "elixir-alice"]
+             [:name "Alice Applicant"]}
+           (set (rename-userid-attributes {:old_sub "alice"
+                                           :sub "elixir-alice"
+                                           :name "Alice Applicant"}))))))
+
+(defn find-or-create-user! [id-data]
+  (let [userid-attrs (get-userid-attributes id-data)]
+    (or (some #(user-mappings/get-user-mapping (first %) (second %)) userid-attrs)
+        (find-first users/user-exists? (vals userid-attrs))
+        (let [userid (get-userid id-data)]
+          (users/add-user-raw! userid id-data)
+          userid))))
+
+(defn save-user-mappings! [id-data userid]
+  (let [attrs (->> (getx env :oidc-userid-attributes)
+                   (map #(or (:rename %) (:attribute %)))
+                   (map keyword))]
+    (doseq [[attr value] (select-keys id-data attrs)
+            :when (not= value userid)]
```

Comment:
not sure about if the `:when` is needed - i'm thinking that those mappings are somewhat redundant:

```clojure
{:userid "123"
 :ext-id-attribute "xyzzy"
 :ext-id-value "123"}
```

Reply:
I think the motivation to include these plain mappings would be to store that the `userid` was from `xyzzy` sometime.